onnxruntime_training_go.go

```diff
@@ -0,0 +1,531 @@
+// Note that currently the final model can only be exported if the session has been
+// initialized with NewTrainingSession and the path to the eval model has been provided.
+func (s *TrainingSession) ExportModel(path string, outputNames []string) error {
+	if s.evalModelPath == nil {
+		return fmt.Errorf(`final model can only be exported if the eval model path is 
+		provided at session creation time (see NewTrainingSession)`)
+	}
+	if path == "" {
+		return fmt.Errorf("Path cannot be empty")
+	}
+	dir, _ := filepath.Split(path)
+	if _, err := os.Stat(dir); dir != "" && os.IsNotExist(err) {
+		return fmt.Errorf("Directory %s does not exist", dir)
+	}
+
+	cOutputNames := make([]*C.char, len(outputNames))
+	for i, name := range outputNames {
+		cOutputNames[i] = C.CString(name)
+	}
+	cPath := C.CString(path)
+	outputLength := C.size_t(len(outputNames))
+	defer func() {
+		for i := range cOutputNames {
+			C.free(unsafe.Pointer(cOutputNames[i]))
+		}
+		C.free(unsafe.Pointer(cPath))
+	}()
+	status := C.ExportModel(s.ortTrainingSession, cPath, outputLength, &cOutputNames[0])
+	if status != nil {
+		return statusToError(status)
+	}
+	return nil
+}
+
+// SaveCheckpoint can be used to save the current checkpoint state at the specified path.
+// This is useful to snapshot the training parameters to continue training later or on
+// a different machine.
+func (s *TrainingSession) SaveCheckpoint(path string, saveOptimizerState bool) error {
+	if path == "" {
+		return fmt.Errorf("Path cannot be empty")
+	}
+	dir, _ := filepath.Split(path)
+	if _, err := os.Stat(dir); dir != "" && os.IsNotExist(err) {
+		return fmt.Errorf("Directory %s does not exist", dir)
+	}
+
+	cPath := C.CString(path)
+	var saveOptimizer int
+	if saveOptimizerState {
+		saveOptimizer = 1
+	}
+
+	defer func() {
+		C.free(unsafe.Pointer(cPath))
+	}()
+
+	status := C.SaveCheckpoint(s.ortCheckpointState, cPath, C.size_t(saveOptimizer))
+	if status != nil {
+		return statusToError(status)
+	}
+	return nil
+}
+
+// Destroy frees all the C memory associated to a training session.
+func (s *TrainingSession) Destroy() error {
+	if s.ortTrainingSession != nil {
+		C.ReleaseOrtTrainingSession(s.ortTrainingSession)
+		s.ortTrainingSession = nil
+	}
+	// note: checkpoint MUST be released after session
+	if s.ortCheckpointState != nil {
+		C.ReleaseCheckpointState(s.ortCheckpointState)
+		s.ortCheckpointState = nil
+	}
+	C.free(unsafe.Pointer(s.trainingModelPath))
+	s.trainingModelPath = nil
+	C.free(unsafe.Pointer(s.evalModelPath))
+	s.evalModelPath = nil
+	C.free(unsafe.Pointer(s.optimizerModelPath))
+	s.optimizerModelPath = nil
+	s.inputs = nil
+	s.outputs = nil
+	return nil
+}
+
+// TrainStep performs the training step.
+func (s *TrainingSession) TrainStep() error {
+	inputLength := C.size_t(len(s.inputs))
+	outputLength := C.size_t(len(s.outputs))
+	status := C.TrainStep(s.ortTrainingSession, inputLength, &s.inputs[0], outputLength, &s.outputs[0])
```

Comment:
Do you make sure that inputs and outputs are have a non-zero length somewhere?  If not, doing `&s.inputs[0]` will cause a panic if the length is 0, regardless of the `inputLength` arguments.  (I had this same issue in the non-training API.)  Preferably, just ensure that `inputs` and `outputs` have nonzero lengths when creating the `TrainingSession`.  (I apologize if you already check this somewhere and I missed it!)